File: Services/ServiceMessenger/app/src/main/java/com/alex/servicemessenger/MainActivity.java
package com.alex.servicemessenger;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;

public class MainActivity extends AppCompatActivity {

  /**
   * Messenger for communicating with the service.
   */
  Messenger service = null;

  /**
   * Flag indicating whether we have called bind on the service.
   */
  boolean isBound;

  /**
   * Class for interacting with the main interface of the service.
   */
  private ServiceConnection connection = new ServiceConnection() {
    public void onServiceConnected(ComponentName className, IBinder service) {
      // This is called when the connection with the service has been
      // established, giving us the object we can use to
      // interact with the service.  We are communicating with the
      // service using a Messenger, so here we get a client-side
      // representation of that from the raw IBinder object.
      MainActivity.this.service = new Messenger(service);
      isBound = true;
    }

    public void onServiceDisconnected(ComponentName className) {
      // This is called when the connection with the service has been
      // unexpectedly disconnected -- that is, its process crashed.
      service = null;
      isBound = false;
    }
  };

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
  }

  public void sayHello(View v) {
    if (!isBound) {
      return;
    }
    // Create and send a message to the service, using a supported 'what' value
    Message msg = Message.obtain(null, MessengerService.MSG_SAY_HELLO, 0, 0);
    try {
      service.send(msg);
    } catch (RemoteException e) {
      e.printStackTrace();
    }
  }

  @Override
  protected void onStart() {
    super.onStart();
    // Bind to the service
    bindService(new Intent(this, MessengerService.class), connection,
        Context.BIND_AUTO_CREATE);
  }

  @Override
  protected void onStop() {
    super.onStop();
    // Unbind from the service
    if (isBound) {
      unbindService(connection);
      isBound = false;
    }
  }
}

File: Services/MessengerServiceClient/app/src/main/java/ru/abaryzhikov/messengerserviceclient/MainActivity.kt
package ru.abaryzhikov.messengerserviceclient

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.*
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private var serviceMessenger: Messenger? = null
    private var bound: Boolean = false
    private val connection = MessengerServiceConnection()
    private val incomingMessenger = Messenger(IncomingHandler { callbackView.text = it })

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        askButton.setOnClickListener { askTimestamp() }
        closeButton.setOnClickListener { this@MainActivity.finish() }
        bindButton.setOnClickListener { doBind() }
    }

    private fun doBind() {
        val intent = Intent().apply {
            setClassName(
                "ru.abaryzhikov.messengerservice",
                "ru.abaryzhikov.messengerservice.MessengerService"
            )
        }
        bindService(intent, connection, Context.BIND_AUTO_CREATE)
    }

    override fun onStop() {
        super.onStop()
        // Unbind from the service.
        if (bound) {
            unbindService(connection)
            bound = false
            callbackView.text = getString(R.string.disconnected)
        }
    }

    private fun askTimestamp() {
        if (!bound) return
        val msg: Message = Message.obtain(null, MSG_GET_TIMESTAMP, 0, 0)
        msg.replyTo = incomingMessenger
        try {
            serviceMessenger?.send(msg)
        } catch (e: RemoteException) {
            Log.e(TAG, "Can't send message", e)
        }
    }

    class IncomingHandler(private val receiver: (String?) -> Unit) : Handler() {
        override fun handleMessage(msg: Message?) {
            when (msg?.what) {
                MSG_GET_TIMESTAMP -> receiver((msg.obj as Bundle).getString(TIMESTAMP_KEY))
                else -> super.handleMessage(msg)
            }
        }
    }

    inner class MessengerServiceConnection : ServiceConnection {

        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            serviceMessenger = Messenger(service)
            bound = true
            callbackView.text = getString(R.string.connected)
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            serviceMessenger = null
            bound = false
            callbackView.text = getString(R.string.disconnected)
        }
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val MSG_GET_TIMESTAMP = 1
        private const val TIMESTAMP_KEY = "Timestamp"
    }
}
